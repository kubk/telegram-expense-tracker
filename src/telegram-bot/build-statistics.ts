import { BankAccount, Currency } from '@prisma/client';
import { Markup } from 'telegraf';
import { BotCallbackQuery } from './bot-action';
import {
  type UserTransactionExpenseRowItem,
  UserTransactionListFilter,
} from '../repository/transaction-repository';
import { formatMoney } from './format-money';
import { UnreachableCaseError } from 'ts-essentials';

type StatisticGroupByType = 'week' | 'month';

const getInlineQueryType = (type: StatisticGroupByType) => {
  switch (type) {
    case 'month':
      return BotCallbackQuery.SelectStatisticsMonth;
    case 'week':
      return BotCallbackQuery.SelectStatisticsWeek;
    default:
      throw new UnreachableCaseError(type);
  }
};

const buildStatisticGrid = (
  row: UserTransactionExpenseRowItem,
  type: StatisticGroupByType,
  bankAccount: { id: string; currency: Currency }
) => {
  const inlineQueryType = getInlineQueryType(type);

  return [
    Markup.button.callback(
      `${row.groupname} ${formatMoney(
        row.difference,
        bankAccount.currency
      ).padStart(10, ' ')}`,
      `${inlineQueryType}:${bankAccount.id}:${row.groupyear}:${row.groupnumber}:${UserTransactionListFilter.NoFilter}`
    ),
    Markup.button.callback(
      formatMoney(row.income, bankAccount.currency),
      `${inlineQueryType}:${bankAccount.id}:${row.groupyear}:${row.groupnumber}:${UserTransactionListFilter.OnlyIncome}`
    ),
    Markup.button.callback(
      formatMoney(row.outcome, bankAccount.currency),
      `${inlineQueryType}:${bankAccount.id}:${row.groupyear}:${row.groupnumber}:${UserTransactionListFilter.OnlyOutcome}`
    ),
  ];
};

export const buildMonthStatistics = (
  statisticRows: UserTransactionExpenseRowItem[],
  bankAccount: { id: string; currency: Currency }
) => {
  return [
    ...statisticRows.map((row) => {
      return buildStatisticGrid(row, 'month', bankAccount);
    }),
    [
      Markup.button.callback(
        '◀️ Back',
        `${BotCallbackQuery.SelectBankAccount}:${bankAccount.id}`
      ),
    ],
  ];
};

export const buildWeekStatistics = (
  statisticRows: UserTransactionExpenseRowItem[],
  bankAccount: BankAccount
) => {
  return [
    ...statisticRows.map((row) => {
      return buildStatisticGrid(row, 'week', bankAccount);
    }),
    [
      Markup.button.callback(
        '◀️ Back',
        `${BotCallbackQuery.SelectBankAccount}:${bankAccount.id}`
      ),
    ],
  ];
};
