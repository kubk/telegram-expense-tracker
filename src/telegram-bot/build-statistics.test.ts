import { buildMonthStatistics } from './build-statistics';
import { Currency } from '@prisma/client';

test('build statistics', () => {
  expect(
    buildMonthStatistics(
      [
        {
          income: 1 * 100,
          currency: Currency.TRY,
          outcome: -10 * 100,
          groupname: 'Apr',
          groupyear: 2022,
          groupnumber: 2,
          difference: -9 * 100,
        },
        {
          income: 1000 * 100,
          currency: Currency.TRY,
          outcome: -999 * 100,
          groupyear: 2022,
          groupname: 'Mar',
          groupnumber: 1,
          difference: 1 * 100,
        },
      ],
      {
        id: '1',
        currency: 'TRY',
      }
    )
  ).toMatchSnapshot();
});
