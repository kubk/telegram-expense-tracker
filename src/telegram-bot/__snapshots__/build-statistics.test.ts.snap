// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build statistics 1`] = `2`;
